### DISCUSSION TITLE
line break

### DISCUSSION THREAD

**Discussion by AntistressStore** [opcodesio/log-viewer#258] [Q&A]
Hello. **Thanks for the great product**. There is an important question that does not allow you to use it comfortably. Line breaks are not interpreted correctly. Inside blocks, pre\n is not interpreted as a line break. And just printed as a symbol. 

And most importantly. If the 'context' element has no more than 110 elements, they are displayed line-by-line. 
![Screenshot_20230815_124308](https://github.com/opcodesio/log-viewer/assets/80528523/80dcf115-6511-40ec-893b-d878ad3c3d3c)

If more, then in solid text, which is very inconvenient. 
![Screenshot_20230815_124434](https://github.com/opcodesio/log-viewer/assets/80528523/611bf374-a569-4660-aee9-7054c210f73a)

Installation 
white-space: pre-line;
or
white-space: pre-wrap;
does not change the situation
**Someone has come across it. How can you fix it?**

**Comment by arukompas** [comment]
Hey @AntistressStore which version of Log Viewer are you using? Line breaks have been improved in latest v2 or in v3 versions.

**Comment by AntistressStore** [comment]
v2.5.3 after your message, I updated the version to the latest 3. And the result is the same

**Reply by arukompas** [reply]
Could you share your operating system, PHP version, and Laravel version as well?

It would also be super useful if you uploaded a log file (as an attachment, not copy/paste text), in case it's a different encoding that Log Viewer doesn't handle well.

**Reply by AntistressStore** [reply]
Good result. If the 'context' element has **no more than 110 elements**, they are displayed line-by-line.
[line-2023-08-16.log](https://github.com/opcodesio/log-viewer/files/12360239/line-2023-08-16.log)

**Reply by AntistressStore** [reply]
Bad result. If more, then in solid text, which is very inconvenient. More than 110 elements
[line-2023-08-16-1.log](https://github.com/opcodesio/log-viewer/files/12360331/line-2023-08-16-1.log)

**Reply by AntistressStore** [reply]
OS: Manjaro Linux x86_64
Kernel: 6.4.9-1-MANJARO
Locale: ru_RU.UTF-8
Laravel Framework 10.19.0
PHP 8.2.9 (cli) (built: Aug  1 2023 12:25:51) (NTS)
Zend Engine v4.2.9, Copyright (c) Zend Technologies
    with Zend OPcache v8.2.9, Copyright (c), by Zend Technologies
    with Xdebug v3.2.2, Copyright (c) 2002-2023, by Derick Rethans

**Reply by arukompas** [reply]
hey @AntistressStore , could you try the latest v2 or v3 please? I've applied a fix on both.

**Reply by gidhublan** [reply]
I am experiencing the same issue. use version "opcodesio/log-viewer": "^3.0", PHP 8.1 , Version: v3.0.4

[laravel.log](https://github.com/opcodesio/log-viewer/files/12408020/laravel.log)

**Reply by arukompas** [reply]
hey @gidhublan , your log seems to work just fine. Make sure to enable `'strip_extracted_context' => true,` in log viewer config so that extracted JSON would also be removed from the message.
Here's what I see on latest Log Viewer version (v3.0.4)

<img width="1091" alt="Screenshot 2023-08-23 at 08 01 51" src="https://github.com/opcodesio/log-viewer/assets/8697942/3f914fb7-2d56-4968-8d53-eb444fcbefd1">

--

If `'strip_extracted_context'` is false, then you would see something like this:

<img width="1410" alt="Screenshot 2023-08-23 at 08 05 23" src="https://github.com/opcodesio/log-viewer/assets/8697942/1d8c2d6c-99db-43e2-acf9-17d4e2cd9838">

**Reply by gidhublan** [reply]
I uninstalled and reinstalled everything and cleared the cache. Currently, there seems to be no issue. However, in the old logs, the same problem continues. The 'strip_extracted_context' is set to true. By the way, thank you very much. It's a valuable component. Thanks again.

**Reply by AntistressStore** [reply]
In my logs everything is the same as before.  For large texts - solid text.  The setting mentioned above does not change anything.  By the way, the new version after updating, deleting and reinstalling did not start in my firefox.  There was an error about the unexpected termination of the script.  Everything was fine on other browsers.  Now it works in firefox too.  There may have been a problem with the browser cache.  I want to understand the problem, because I like your product, it is very convenient.  But the continuous text of errors is extremely difficult to read.  If I have time, I'm ready to help in development, if you tell me where it is better to look in the code so that I don't search all over the code.

**Reply by AntistressStore** [reply]
Perhaps I need to arrange my logs in a special way so that the parser correctly parses the context?  How does the parser understand that it needs to parse the context line by line?

**Reply by arukompas** [reply]
Hey @AntistressStore , the best place to start understanding how the Laravel log is parsed is the `Opcodes\LogViewer\Logs\LaravelLog` class. It receives a large block of text that represents a single log, and then works with regex to extract various elements, such as date, severity, message, context, etc. Have a look at the class and maybe something sticks out.

September will be a difficult month for me to do OSS work, but I'll be back in full force in October and I'll be able to help more then.

If you do find anything, let me know! Any help is very much appreciated 🙏

**Reply by pasamsin** [reply]
I think the issue lies with the regular expression $pattern = '/(\{(?:[^{}]|(?R))*\}|\[(?:[^\[\]]|(?R))*\])/';, where preg_match_all seems to be capturing only the date information '[2024-01-25 15:47:00]'. I've tried the regex on regex websites and it matches, but when used with preg_match_all, it doesn't match.

**Reply by pasamsin** [reply]
If greater than a certain length, preg_match_all will not work on PHP versions above 7.2.

**Reply by pasamsin** [reply]
This could be due to the PCRE2 update. I've increased the values of pcre.backtrack_limit and pcre.recursion_limit, but there has been no change.

**Reply by pasamsin** [reply]
I reorganized the expression for pcre2 as follows. I'm not very knowledgeable about regex, so it might be incorrect, but for now, it seems to be working correctly.

        `$pattern = '/\{(?:[^{}]*+\{[^{}]*+\}[^{}]*)*+\}|\[(?:[^\[\]]*+\[[^\[\]]*+\][^\[\]]*)*+\]/';`

**Reply by pasamsin** [reply]
If the correct regex is written, the issue will be resolved. It's not working due to PCRE2
